### PR TITLE
Add globalThis to polyfill for old browswers

### LINKED ISSUE #45587: App Directory: globalThis is not defined under Chome 71 version
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.2.0: Fri Nov 11 02:04:44 PST 2022; root:xnu-8792.61.2~4/RELEASE_ARM64_T8103
    Binaries:
      Node: 16.19.0
      npm: 8.19.3
      Yarn: 1.22.18
      pnpm: 7.21.0
    Relevant packages:
      next: 13.1.7-canary.5
      eslint-config-next: 13.1.7-canary.5
      react: 18.2.0
      react-dom: 18.2.0

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to the code that reproduces this issue

https://github.com/CrazydduGi/nextjs-app-directory-global-this-reproduction

### To Reproduce

In order to reproduce the bug is just fallow below steps

1. clone reproducion link code
2. yarn build
3. yarn start
4. connect http://localhost:3000 with Chrome which versions is under 71
5. open inspector and then see the console error and javascript stops

Screenshot 2023-02-06 at 2 08 58 AM

### Describe the Bug

According to Nextjs documentation ([Link](https://nextjs.org/docs/basic-features/supported-browsers-features))

Nextjs is support Chrome 64+, however `globalThis` is not supported yet until Chrome 71 version so that `globalThis` is not defined in lower version between 64 and 71

### Expected Behavior

console error does not occur for globalThis access

### Which browser are you using? (if relevant)

Chrome 70

### How are you deploying your application? (if relevant)

next build && next start

### PR DESCRIPTION
**Pull request by CrazydduGi**
fixes #45587
In this [PR](https://github.com/vercel/next.js/pull/45589), @shuding  already tried to fix #45587 bug.
However there are multiple places that using `globalThis` in browser environment.
under `packages/next/src/compiled` there are many places like `react-server-dom-webpack`, `react/cjs`, etc.

So I added globalThis polyfill on `polyfill module`

## Bug

- [x] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [x] Errors have a helpful link attached, see [`contributing.md`](https://github.com/vercel/next.js/blob/canary/contributing.md)

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [x] Related issues linked using `fixes #number`
- [ ] [e2e](https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md#writing-tests-for-nextjs) tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have a helpful link attached, see [`contributing.md`](https://github.com/vercel/next.js/blob/canary/contributing.md)

## Documentation / Examples

- [ ] Make sure the linting passes by running `pnpm build && pnpm lint`
- [ ] The "examples guidelines" are followed from [our contributing doc](https://github.com/vercel/next.js/blob/canary/contributing/examples/adding-examples.md)

### CODE REVIEW

**Review by SukkaW — CHANGES_REQUESTED** [review]
Per Next.js documentation: https://nextjs.org/docs/basic-features/supported-browsers-features

Next.js only includes polyfills to support up to **ES2019**. Currently, no further polyfill should be added in the `next-polyfill-module` as it will result in a client-side bundle size increase.

Since `globalThis` is introduced in **ES2020** which is not prior to ES2019, Next.js should not include this polyfill by default.

If you want to introduce the polyfill for ES2020 features (like `globalThis`), then you should also polyfill every ES2020 feature in order to maintain behavior consistency (to prevent future issues like "why globalThis is polyfilled but `String.prototype.matchAll` is not"):

- globalThis
- ~~BigInt~~ - it is not polyfill-able
- Promise.allSettled()
- String.prorotype.matchAll

See also: https://github.com/vercel/next.js/blob/canary/packages/eslint-plugin-next/src/rules/no-unwanted-polyfillio.ts

File: packages/next-polyfill-module/src/index.js [hunk 1]
```diff
@@ -118,3 +118,17 @@ if (!Object.fromEntries) {
...
+ * Firefox: 65
+ * Chrome: 71
+ * Safari: 12.1
+ *
+ * https://caniuse.com/mdn-javascript_builtins_globalthis
+ */
+// Assume this polyfill module only running in browser environment.
+if (typeof globalThis === 'undefined' && typeof window !== 'undefined') {
+  window.globalThis = window
+}
```
Comment (SukkaW) [inline packages/next-polyfill-module/src/index.js:134]:
```suggestion
if (
  typeof globalThis !== 'object'
  // check corrupted globalThis
  // https://github.com/zloirock/core-js/blob/524af814dc4aef5d05102278ef288dfbd401ec55/packages/core-js/internals/global.js#L2
  || globalThis.Math !== Math
) {
  // use "self" as a fallback in case "window" doesn't exist
  const global = typeof window !== 'undefined' ? window : self;
  global.globalThis = global;
}
```

**Review by SukkaW — CHANGES_REQUESTED** [review]

File: packages/next-polyfill-module/src/index.js (hunk 1 above)
Comment (SukkaW) [reply packages/next-polyfill-module/src/index.js:134]: To prevent `Function('return this')()`, you can also use another approach: https://mathiasbynens.be/notes/globalthis

```suggestion
if (typeof globalThis !== 'object') {
  Object.defineProperty(Object.prototype, '_T_', {
    configurable: true,
    get: function get() {
      var global = this || self;
      global.globalThis = global;
      delete Object.prototype._T_;
    }
  });
  _T_;
);

```

**Comment by mcAnastasiou** [comment]
Do we have any update on this? We are getting many errors in our app regarding "Can't find variable: globalThis"

**Comment by rusted-love** [comment]
@ijjk Is there any reason on this PR not to be approved? It's important for companies to have wide support of devices. More and more websites use `appDir` on production with each passing day.
Workaround to place `import 'core-js/features/global-this'` in  root layout does not work as in `pages` `_app.tsx`.

**Comment by maiconcarraro** [comment]
> Per Next.js documentation: https://nextjs.org/docs/basic-features/supported-browsers-features
> 
> Next.js only includes polyfills to support up to **ES2019**. Currently, no further polyfill should be added in the `next-polyfill-module` as it will result in a client-side bundle size increase.

hey @SukkaW, I think the confusion is that on supported browsers it says "Chrome 64+", but ES2019 was released on [Chrome 71+](https://caniuse.com/?search=globalThis), or at least around that.

About the bundle size, I agree that It shouldn't be in scope of default behavior, I gave a try with `experimental.legacyBrowsers` and it doesn't match the polyfill for Chrome 64+ as well.

I have an emulator using Android 9 with default Chrome 69 and it doesn't work, once I update to Chrome 114 it works fine.

Let me know if I'm missing something, the problem is not the poylfill but the supported list.

**Comment by nxl3477** [comment]
add polyfill script on app/layout.tsx


 <html lang="en">
      <head>
        {/* < chrome 71 */}
        <script src="https://polyfill.io/v3/polyfill.min.js?features=globalThis"></script>
      </head>
      <body>
        <main>
          {children}
        </main>
      </body>
    </html>

**Comment by rusted-love** [comment]
This solution helped. 
`<script
					dangerouslySetInnerHTML={{
						__html: 'window.globalThis = window',
					}}
					type="text/javascript"
				/>`

**Comment by FeiPengDev** [comment]
Could you please tell me if there is any progress on this issue? Because I didn’t see this pr being merged.

**Comment by liucan233** [comment]
> Could you please tell me if there is any progress on this issue? Because I didn’t see this pr being merged.

swc会polyfillglobalThis，但是nextjs实际运行没有，估计是nextjs swc配置问题，短时间内解决不了这个问题，有时间我会看看这个问题。

**Comment by SukkaW** [comment]
> swc会polyfillglobalThis，但是nextjs实际运行没有，估计是nextjs swc配置问题，短时间内解决不了这个问题，有时间我会看看这个问题。

Next.js doesn't rely on swc's `mode: "usage"` to create polyfills:

- Both Babel and swc **will always** introduce **unnecessary polyfills** to the final bundle, wasting disk space and bandwidth.
  - E.g. if you write `something.at(-1)`, you will have all [`es.array.at`](https://github.com/zloirock/core-js/blob/master/packages/core-js/modules/es.array.at.js), [`es.string.at-alternative`](https://github.com/zloirock/core-js/blob/master/packages/core-js/modules/es.string.at-alternative.js), and [`es.typed-array.at`](https://github.com/zloirock/core-js/blob/master/packages/core-js/modules/es.typed-array.at.js) included in your final bundle.\
  - It is impossible for Babel and swc to fix this issue, because Babel and swc will never know the exact type of `something` inside `something.at(-1)` (Is it string, is it array, or it it a typed array?)
- The `globalThis` polyfill from `core-js` is unsafe since it uses `new Function()` as a fallback. In a secure execution context (Browser with CSP enabled, Node.js with SES lockdown, Cloudflare Workers/Pages Functions, Netlify Functions, Vercel Edge Runtime, etc.) **it will always fail catastrophically**

In short, unless further investigation has been made and the solution has been developed, Next.js **will never** use swc to automatically include polyfills.

**Comment by FeiPengDev** [comment]
> > swc会polyfillglobalThis，但是nextjs实际运行没有，估计是nextjs swc配置问题，短时间内解决不了这个问题，有时间我会看看这个问题。
> 
> Next.js doesn't rely on swc's `mode: "usage"` to automatically include polyfills to the output:
> 
> * Both Babel and swc **will always** introduce **unnecessary polyfills** to the final bundle with `mode: "usage"`, **wasting disk space and bandwidth**.
>   
>   * E.g. if you write `something.at(-1)`, you will have all [`es.array.at`](https://github.com/zloirock/core-js/blob/master/packages/core-js/modules/es.array.at.js), [`es.string.at-alternative`](https://github.com/zloirock/core-js/blob/master/packages/core-js/modules/es.string.at-alternative.js), and [`es.typed-array.at`](https://github.com/zloirock/core-js/blob/master/packages/core-js/modules/es.typed-array.at.js) included in your final bundle.
>   * This behavior is intended, because it is impossible for Babel and swc to identify the exact type of `something` inside `something.at(-1)` (Is it a string, is it an array, or is it a typed array?).
> * The `globalThis` polyfill from `core-js` is unsafe since it uses `new Function()` as a fallback. In a secure execution context (Browser with CSP enabled, Node.js with SES lockdown, Cloudflare Workers/Pages Functions, Netlify Functions, Vercel Edge Runtime, etc.) **it will inevitably fail catastrophically**.
> 
> In short, unless further investigations have been made and the solution has been developed, Next.js **will most likely never** use swc to automatically include polyfills.

So we can only add polyfill ourselves at present, right? 
Like https://github.com/vercel/next.js/pull/45592#issuecomment-1596837885

**Comment by kpervin** [comment]
With @apollo/client v3.8.* this is now an issue we've run into. They use `globalThis.__DEV__` which breaks browser compatability for < Chrome 71 versions.

**Comment by shra1-gzp** [comment]
Any traction on this? We require this fix for some Oppo Browsers. 
Right now, we're relying on this:
`<script dangerouslySetInnerHTML={{ __html: 'window.globalThis = window', }} type="text/javascript" />`

**Comment by kpervin** [comment]
This is also now an issue given that `graphql: ^16.7.1` now uses `globalThis`. At least with Apollo we could use 

```ts
/**
 * @type {import('next').NextConfig}
 */
const config = {
  webpack: (config, { webpack }) => {
    if (process.env.NODE_ENV === "production") {
      config.plugins.push(
        new webpack.DefinePlugin({
          "globalThis.__DEV__": false,
        }),
      );
    }
    return config;
  },
}
```
but attempting to override the entirety of `globalThis` in the webpack config will likely cause cascading effects. It would be good to have a polyfill built into NextJS for these possibilities.

**Comment by laogui** [comment]
In app router mode, adding a script tag works in dev mode, but it doesn't work after build

**Comment by nykula** [comment]
@laogui Indeed, staying on next ^14.1.0 because newer minor versions optimize the build in a way that's incompatible with the WebView of Android 7 kiosk tablets deployed with my app.

**Comment by timneutkens** [comment]
The minimum supported browsers have been updated to `baseline widely available`: https://nextjs.org/docs/app/getting-started/installation#supported-browsers which means that adding this polyfill is no longer needed by default. If you want to support older browsers you'll have to bring your own polyfills in instrumentation-client.js (or .ts)
